### PR TITLE
Refactor/data structures

### PR DESCRIPTION
**Pull request by JeroenMandersloot**
Refactored the data structures used in our application. This PR cannot be merged into master yet, because scripts like `run.py` still depend on the old data structures. 

### Datatypes
I removed `lr_face/data_providers.py` completely, so classes like `ImageWithIDs`, `PairWithIDs` and `ImagePairs` no longer exist. These have been dropped in favor of conceptually simpler data classes:
- `FaceImage`
- `FacePair`
- `FaceTriplet`

The most basic data element is a `FaceImage`, which is simply a convenient way of storing and working with the image data we use throughout our application. In it, the `path` to the image file and a unique `identity` of the person depicted on the image are stored. Additionally, it implements a `get_image()` method to return the actual pixel data as a `np.ndarray`. In the future we should also add a `get_embedding(model)` method that computes and returns the embedding of the image given the specified `model`.

By passing around `FaceImage` instances it is always clear throughout the application what kind of data we're working with. For convenience I also created two wrapper classes for two other commonly used data structures: `FacePair` and `FaceTriplet`. `FacePair` is nothing more than a thin wrapper around two `FaceImage` instances with a convenient `same_identity` property to find out whether the two images depict the same person. Similarly, `FaceTriplet` is a thin wrapper around three `FaceImage` instances that enforces that the `anchor` and `positive` image share the same identity while the `anchor` and `negative` image do not.

### Datasets

I rewrote the way we handle datasets to move from functions to classes. This allows us to get rid of things like `ImageProvider` and `PairProvider` so we no longer have to think about whether a particular dataset returns images or pairs, and how to handle both cases. Now, every dataset is in essence a wrapper around `List[FaceImage]`, i.e. it handles a collection of images.

I implemented an abstract `Dataset` class that implements most of the functionality shared between all datasets. Each concrete dataset, such as the ENFSI dataset for example, inherits from `Dataset` and has to implement only a single method: `images()`, which is actually a property that loads and returns a list of `FaceImage` instances for that particular dataset. Each dataset has default logic for converting this list of images into a list of `FacePair` or `FaceTriplet` instances through the `pairs` and `triplets` properties. For datasets that define a specific set of pairs or triplets, these properties can be overridden in the subclass.

### Helper functions

There are three simple helper functions that make dealing with the new data structures even simpler. The `make_pairs()` method takes a list of `FaceImage` instances and pairs them up to return a list of `FacePair` instances. Pairing can be done in one of several configurable ways (e.g. only make positive/negative pairs, make both positive and negative pairs, etc.). The `make_triplets()` function works in much the same way, except it returns a list of `FaceTriplet` instances.

Then there is the `to_array()` method, which converts a list of `FaceImage` instances to the appropriate representation as an `np.ndarray`, which can then be used in ML pipelines. This way, we don't have to pass along numpy arrays all the time, which would make the code very hard to follow. Instead, you can pass along `Dataset` objects or lists of `FaceImage` instances and convert those to numpy arrays only when you need to.

Finally I implemented a new `@cache` decorator which allows you to cache the result of a method call that involves heavy computations (such as loading an entire dataset) and immediately return the same result the next time the method is called with the same arguments. To use it, simply add `@cache` as a decorator above your method and it will work! Two caveats:
- All the arguments to the method should be hashable
- If the method also has a `@property` decorator, be sure that apply the `@property` decorator last. This is not something specific to the `@cache` decorator, but `@property` decorators always have to be applied last because they return a `property` object instead of a function.

### Visualization

I've tried to visualize these new data structures and their dependencies on each other below:

datastructures

### CODE REVIEW

File: tests/test_data.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+###############
+
+@skip_if_missing(ForenFaceDataset)
+def test_forenface_dataset_has_correct_num_images(forenface):
+    assert len(forenface.images) == 2476
+
+
+@skip_if_missing(ForenFaceDataset)
+def test_forenface_dataset_has_correct_num_pairs(forenface):
+    assert len(forenface.pairs) == 60798
```
Comment (rolfypma) [inline tests/test_data.py:193]: wil je deze hard hierin zetten? Er zijn geen natuurlijke paren toch?

File: lr_face/data.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+    handle annotated images in a unified way. All datasets should preferably
+    be a wrapper around handling lists of `FaceImage` instances.
+    """
+
+    # The path to the image file.
+    path: str
+
+    # A globally unique identifier for the person depicted on the image, only
+    # shared with other images that depict the same person.
+    identity: str
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:29]: Nu niet nee, ik zou zo gauw ook niet weten hoe je dat moet doen :thinking: Want twee afbeeldingen hebben dus wel dezelfde `identity` als ze dezelfde persoon afbeelden. Heb je daar ideeen over?

File: lr_face/data.py [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+        :param normalize: bool
+        :return: np.ndarray
+        """
+        res = cv2.imread(self.path)
+        if res is None:
+            raise ValueError(f'Reading {self.path} resulted in None')
+        if res.shape[-1] != 3:
+            raise ValueError(f'Expected 3 channels, got {res.shape[-1]}')
+        if resolution:
+            res = cv2.resize(res, resolution)
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:56]: Ja ik zou willen proberen de resize logica zoveel mogelijk op 1 plek te hebben, dus dan zorgen dat de andere plekken waar dit gebeurt deze methode gaan aanroepen.

[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:56]: - [ ] Preprocessing script resizen updaten

File: lr_face/data.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+                data.append(FaceImage(image_path, identity, {
+                    'source': str(self)
+                }))
+        return data
+
+    @property
+    @cache
+    def pairs(self) -> List[FacePair]:
+        pairs = []
+        with open(os.path.join(self.ROOT, 'pairs.txt'), 'r') as f:
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:282]: Misschien niet, de enige reden die ik zo gauw kon bedenken was dat papers misschien hun resultaten posten op basis van de resultaten op deze specifieke pairs ipv de hele dataset. Maar als dat niet het geval is, hebben we dit in principe niet nodig nee.

File: tests/test_data.py (hunk 1 above)
Comment (JeroenMandersloot) [reply tests/test_data.py:193]: Goed punt ja, ik haal deze eruit

- [ ] Remove unit test for pairs in forenface

File: lr_face/data.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+class TestDataset(Dataset):
+    @property
+    @cache
+    def images(self) -> List[FaceImage]:
+        ids = [1, 1, 1, 2, 2, 2, 3, 3, 4, 4, 5]
+        return [DummyFaceImage(path='', identity=f'TEST-{idx}') for idx in ids]
+
+
+class ForenFaceDataset(Dataset):
+    ROOT = os.path.join('resources', 'forenface')
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:239]: - [ ] Rename `ROOT` naar `RESOURCE_FOLDER`

File: lr_face/data.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+        Uses the specified `model` to compute an embedding of this image.
+        Depending on the model, the dimensionality of the embedding may differ.
+        Returns a 1D array of shape `(embedding_size)`.
+
+        TODO: implement.
+        TODO: implement hashing in `BaseModel`, otherwise caching doesn't work.
+        """
+        raise NotImplementedError
+
+    def __post_init__(self):
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/data.py:73]: Haha, is geen algemene methode, maar bestaat specifiek voor dataclasses om dit soort validaties en zo te doen :)

File: lr_face/utils.py [hunk 7]
```diff
@@ -194,3 +196,57 @@ def fix_tensorflow_rtx():
     gpu_devices = tf.config.experimental.list_physical_devices('GPU')
     for device in gpu_devices:
         tf.config.experimental.set_memory_growth(device, True)
+
+
+def cache(func):
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/utils.py:201]: `@lru_cache` had ik even naar gekeken maar die wilde altijd een maxsize argument wat een beetje irritant is. En deze eigen implementatie van `@cache` had ik nog ergens liggen :sweat_smile: Maar we zouden wel `@cache` gewoon een alias voor `@lru_cache(maxsize=None)` kunnen maken denk ik :thinking:

File: tests/src/util.py [hunk 8]
```diff
@@ -1,8 +1,21 @@
...
 
+def get_project_path(relative_path: str) -> str:
+    """
+    Takes a path relative to the root directory of the project (the directory
+    above the `tests` directory) and returns an absolute path to it.
+    
+    :param relative_path: str
+    :return: str
+    """
+    root_path = dirname(dirname(dirname(abspath(__file__))))
```
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply tests/src/util.py:15]: We merken het snel genoeg als dit ooit fout gaat, dan falen als het goed is een heleboel testen in 1x, dus hij wordt automatisch meegetest

File: lr_face/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rolfypma) [reply lr_face/utils.py:201]: ben wel meer van standaardfuncties gebruiken dan eigen code, als het kan

File: lr_face/data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rolfypma) [reply lr_face/data.py:73]: weer wat geleerd:)

File: lr_face/data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rolfypma) [reply lr_face/data.py:29]: kwam niet verder dan datasetname prependen zodat daar geen clash kan ontstaan, maar dat had je al:)

File: lr_face/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JeroenMandersloot) [reply lr_face/utils.py:201]: - [ ] `@cache` gebruik laten maken van `@lru_cache`
